File: src/types/copilot.ts
export type KnowledgeRef = {
  sourceId: string
  title: string
  snippet: string
  score: number
}

export type AiActionKind = 'navigate' | 'generate' | 'open'

export type AiAction = {
  id: string
  label: string
  kind: AiActionKind
  payload: string
}

export type RoadmapArtifact = {
  type: 'roadmap'
  title: string
  phases: { name: string; duration: string; focus: string; progress: number }[]
}

export type SprintArtifact = {
  type: 'sprint'
  title: string
  metrics: { label: string; value: string }[]
  highlights: string[]
  risks: string[]
}

export type RiskArtifact = {
  type: 'risk'
  title: string
  items: { name: string; severity: 'high' | 'medium' | 'low'; mitigation: string }[]
}

export type Artifact = RoadmapArtifact | SprintArtifact | RiskArtifact

export type AiMessage = {
  id: string
  role: 'user' | 'assistant'
  content: string
  timestamp: string
  citations?: KnowledgeRef[]
  actions?: AiAction[]
  artifact?: Artifact
  contextUsed?: string[]
}

export type AiThread = {
  id: string
  title: string
  createdAt: string
  updatedAt: string
  messages: AiMessage[]
}

File: backend/app/schemas/copilot.py
from typing import List, Literal, Optional

from pydantic import BaseModel

from app.services.ai_copilot import EngineContext


class CopilotChatRequest(BaseModel):
    prompt: str
    threadId: Optional[str] = None
    context: Optional[EngineContext] = None
    consentAccepted: bool = True


class DocumentChatRequest(BaseModel):
    title: str
    content: str
    threadId: Optional[str] = None
    context: Optional[EngineContext] = None
    consentAccepted: bool = False


class Citation(BaseModel):
    sourceId: str
    title: str
    snippet: str
    score: float


class AiAction(BaseModel):
    id: str
    label: str
    kind: Literal["navigate", "generate", "open"] = "navigate"
    payload: str


class CopilotChatResponse(BaseModel):
    content: str
    threadId: str
    messageId: str
    citations: List[Citation] = []
    actions: List[AiAction] = []
    contextUsed: List[str] = []
    model: Optional[str] = None
    mock: bool = False


class ThreadMessageResponse(BaseModel):
    id: str
    role: str
    content: str
    citations: list = []
    createdAt: str


class ThreadResponse(BaseModel):
    id: str
    title: str
    createdAt: str
    updatedAt: str
    messages: List[ThreadMessageResponse]
